sys/include/net/sock/dtls.h

```diff
@@ -0,0 +1,645 @@
+ *     return -1;
+ * }
+ *
+ * if (sock_dtls_session_create(&dtls_sock, &remote, &session) < 0) {
+ *     printf("Error creating session: %d\n", (int)res);
+ *     sock_dtls_close(&dtls_sock);
+ *     sock_udp_close(&udp_sock);
+ *     return -1;
+ * }
+ *
+ * const char data[] = "HELLO";
+ * int res = sock_dtls_send(&dtls_sock, &session, data, sizeof(data));
+ * if (res >= 0) {
+ *     printf("Sent %d bytes: %*.s\n", res, res, data);
+ *     res = sock_dtls_recv(&dtls_sock, &session, rcv, sizeof(rcv), SOCK_NO_TIMEOUT);
+ *     if (res > 0) {
+ *         printf("Received %d bytes: %*.s\n", res, res, rcv);
+ *     }
+ * }
+ * else {
+ *     puts("Error sending data");
+ * }
+ *
+ * sock_dtls_session_destroy(&dtls_sock, &session);
+ * sock_dtls_close(&dtls_sock);
+ * sock_udp_close(&udp_sock);
+ * return 0;
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * @{
+ *
+ * @file
+ * @brief   DTLS sock definitions
+ *
+ * @author  Aiman Ismail <muhammadaimanbin.ismail@haw-hamburg.de>
+ * @author  Martine Lenders <m.lenders@fu-berlin.de>
+ * @author  Raul A. Fuentes Samaniego <raul.fuentes-samaniego@inria.fr>
+ * @author  Daniele Lacamera <daniele@wolfssl.com>
+ * @author  Ken Bannister <kb2ma@runbox.com>
+ */
+
+#ifndef NET_SOCK_DTLS_H
+#define NET_SOCK_DTLS_H
+
+#include <assert.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <sys/types.h>
+
+#include "net/sock.h"
+#include "net/sock/udp.h"
+#include "net/credman.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @name DTLS version number
+ * @anchor sock_dtls_prot_version
+ * @{
+ */
+#define SOCK_DTLS_1_0    (1)    /**< DTLS version 1.0 */
+#define SOCK_DTLS_1_2    (2)    /**< DTLS version 1.2 */
+#define SOCK_DTLS_1_3    (3)    /**< DTLS version 1.3 */
+/** @} */
+
+/**
+ * @name DTLS role
+ * @anchor sock_dtls_role
+ * @{
+ */
+#define SOCK_DTLS_CLIENT (1)    /**< Endpoint client role */
+#define SOCK_DTLS_SERVER (2)    /**< Endpoint server role */
+/** @} */
+
+/**
+ * @brief   Type for a DTLS sock object
+ *
+ * @note    API implementors: `struct sock_dtls` needs to be defined by
+ *          an implementation-specific `sock_dtls_types.h`.
+ */
+typedef struct sock_dtls sock_dtls_t;
+
+/**
+ * @brief Information about a created session.
+ */
+typedef struct sock_dtls_session sock_dtls_session_t;
+
+/**
+ * @brief Called exactly once during autoinit.
+ *
+ * Calls the initialization function required by the DTLS stack used.
+ */
+void sock_dtls_init(void);
+
+/**
+ * @brief Creates a new DTLS sock object
+ *
+ * Takes an initialized UDP sock and uses it for the transport.
+ * Memory allocation functions required by the underlying DTLS
+ * stack can be called in this function.
+ *
+ * @see net_credman.
+ *
+ * @param[out] sock     The resulting DTLS sock object
+ * @param[in] udp_sock  Existing UDP sock initialized with
+ *                      @ref sock_udp_create()to be used underneath.
+ * @param[in] tag       Credential tag of the sock. The sock will only use
```

Comment:
Which sock? To make it clearer that the parameter `sock` is meant, use `@p sock`

Reply:
Fixed.